test/test_producer.py

```diff
@@ -34,10 +34,15 @@ def test_end_to_end(kafka_broker, compression):
         elif platform.python_implementation() == 'PyPy':
             return
 
+    if compression == 'zstd' and env_kafka_version() < (2, 1, 0):
+        return
```

Comment:
This should be using a `pytest` skip... http://doc.pytest.org/en/latest/skipping.html#skipping-test-functions
Could be done inside the test function, but IMO the decorator pattern is more readable when skimming the tests... since decorators fire as the inner ones first, I'd use the following order:
```
@pytest.mark.skipif(compression == 'zstd' and env_kafka_version() < (2, 1, 0), reason="zstd support requires Kafka 2.1 (KIP-110)")
@pytest.mark.skipif(not env_kafka_version(), reason="No KAFKA_VERSION set")
@pytest.mark.parametrize("compression", [None, 'gzip', 'snappy', 'lz4', 'zstd'])
```

Reply:
> This should be using a `pytest` skip... http://doc.pytest.org/en/latest/skipping.html#skipping-test-functions
> Could be done inside the test function, but IMO the decorator pattern is more readable when skimming the tests... since decorators fire as the inner ones first, I'd use the following order:
> 
> ```
> @pytest.mark.skipif(compression == 'zstd' and env_kafka_version() < (2, 1, 0), reason="zstd support requires Kafka 2.1 (KIP-110)")
> @pytest.mark.skipif(not env_kafka_version(), reason="No KAFKA_VERSION set")
> @pytest.mark.parametrize("compression", [None, 'gzip', 'snappy', 'lz4', 'zstd'])
> ```

The compression param is injected by one of the decorators though, so that notation will not work, but i'll give the doc a read and figure out if i can manipulate them after the `parametrize` decorator is in place

Reply:
oh interesting... I was expecting it to wrap the function in a way that made it possible to introspect it... essentially each decorator gives a wrapped function to the one above it.

Instead of a decorator, it's also possible to use `pytest.skip("reason")` function within a test... use that in place of your `return` statement. More details in http://doc.pytest.org/en/latest/skipping.html#skipping-test-functions